File: 프로그래머스/unrated/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.py
def solution(n):
    A = [[0]*n for _ in range(n)]
    r, c, dr, dc = 0, 0, (0,1,0,-1), (1,0,-1,0)

    i = 1
    for m in range(n-1,0,-2):
        for j in range(4):
            for _ in range(m):
                A[r][c] = i
                r, c = r + dr[j], c + dc[j]     
                i += 1
        r, c = r + 1 , c + 1
        
    if n % 2 == 1:
        A[r][c] = n**2
        
    return A

File: 프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.cpp
#include <string>
#include <vector>

using namespace std;

vector<vector<int>> solution(int n) {
    vector<vector<int>> D = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    vector<int> tmp(n, 0);
    vector<vector<int>> answer(n, tmp);
    int r=0, c=-1, d=0, num=1;
    
    for (int i=0; i<2*n-1; i++) {
        for (int j=0; j<n-(i+1)/2; j++) {
            r += D[d][0];
            c += D[d][1];
            answer[r][c] = num;      
            num++;
        }
        d = (d+1)%4;
    }
    return answer;
}